File: packages/flutter-im-ui/test/interaction_state_test.dart
import 'package:flare_im_ui/flare_im_ui.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('composer precedence and allowed actions are deterministic', () {
    expect(
      resolveComposerMode(
        const FlareComposerInteractionInput(
          readOnly: true,
          recording: true,
          online: false,
        ),
      ),
      FlareComposerInteractionMode.readOnly,
    );
    expect(
      resolveComposerMode(
        const FlareComposerInteractionInput(
          permissionGranted: false,
          online: false,
        ),
      ),
      FlareComposerInteractionMode.permissionDenied,
    );
    expect(
      composerAllowedActions(FlareComposerInteractionMode.offline),
      isNot(contains(FlareComposerInteractionAction.send)),
    );
  });

  test('message lifecycle outranks transient hover', () {
    expect(
      resolveMessageMode(
        const FlareMessageLifecycle(
          mutation: FlareMessageMutationState.recalled,
        ),
        hovered: true,
      ),
      FlareMessageInteractionMode.recalled,
    );
    expect(
      resolveMessageMode(
        const FlareMessageLifecycle(read: FlareMessageReadState.read),
      ),
      FlareMessageInteractionMode.read,
    );
  });

  test('selection range, shortcuts, and gesture direction are stable', () {
    final anchor = reduceSelection(
      const FlareSelectionState(),
      const FlareSelectionReplace('b'),
    );
    final range = reduceSelection(
      anchor,
      const FlareSelectionExtend('d', ['a', 'b', 'c', 'd']),
    );
    expect(range.selectedIds, {'b', 'c', 'd'});
    expect(
      resolveDesktopShortcut('k', primary: true),
      FlareDesktopShortcutAction.commandPalette,
    );
    expect(
      resolveSwipeIntent(deltaX: 80, deltaY: 70, message: true),
      FlareSwipeIntent.none,
    );
    expect(
      resolveSwipeIntent(deltaX: 80, deltaY: 8, message: true),
      FlareSwipeIntent.reply,
    );
  });

  test('message capabilities follow ownership and terminal lifecycle', () {
    final own = resolveMessageCapabilities(
      const FlareMessageCapabilityInput(
        lifecycle: FlareMessageLifecycle(),
        own: true,
      ),
    );
    expect(own, contains(FlareMessageCapability.edit));
    expect(own, isNot(contains(FlareMessageCapability.report)));
    final recalled = resolveMessageCapabilities(
      const FlareMessageCapabilityInput(
        lifecycle: FlareMessageLifecycle(
          mutation: FlareMessageMutationState.recalled,
        ),
        own: true,
      ),
    );
    expect(recalled, isEmpty);
  });

  test('extended actions and shortcuts share the capability contract', () {
    final actions = resolveMessageCapabilities(
      const FlareMessageCapabilityInput(
        lifecycle: FlareMessageLifecycle(),
        own: true,
        pinned: true,
        hasThread: true,
        hasQuote: true,
        supportsMergeForward: true,
      ),
    );
    expect(
      actions,
      containsAll({
        FlareMessageCapability.unpin,
        FlareMessageCapability.openThread,
        FlareMessageCapability.jumpToQuote,
        FlareMessageCapability.mergeForward,
      }),
    );
    expect(
      resolveDesktopShortcut(
        'f',
        scope: FlareDesktopShortcutScope.conversation,
      ),
      FlareDesktopShortcutAction.forward,
    );
    expect(
      resolveDesktopShortcut(
        't',
        scope: FlareDesktopShortcutScope.conversation,
      ),
      FlareDesktopShortcutAction.openThread,
    );
  });

  test('action presentation stays downstream of availability', () {
    final actions = resolveMessageActions(
      const FlareMessageCapabilityInput(
        lifecycle: FlareMessageLifecycle(),
        own: false,
      ),
      FlareMessageActionPresentation.hoverToolbar,
    );
    expect(
      actions.singleWhere((a) => a.id == FlareMessageCapability.reply).promoted,
      isTrue,
    );
    expect(
      actions.singleWhere((a) => a.id == FlareMessageCapability.report).group,
      FlareMessageActionGroup.destructive,
    );
  });
}

File: packages/android-im-ui/src/test/kotlin/com/flare/im/ui/InteractionStateTest.kt
package com.flare.im.ui

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Test

class InteractionStateTest {
    @Test fun `composer precedence and actions are deterministic`() {
        assertEquals(FlareComposerInteractionMode.ReadOnly, resolveComposerMode(FlareComposerInteractionInput(readOnly = true, recording = true, online = false)))
        assertEquals(FlareComposerInteractionMode.PermissionDenied, resolveComposerMode(FlareComposerInteractionInput(permissionGranted = false, online = false)))
        assertFalse(composerAllowedActions(FlareComposerInteractionMode.Offline).contains(FlareComposerInteractionAction.Send))
    }

    @Test fun `message lifecycle outranks hover`() {
        assertEquals(FlareMessageInteractionMode.Recalled, resolveMessageMode(FlareMessageLifecycle(mutation = FlareMessageMutationState.Recalled), hovered = true))
        assertEquals(FlareMessageInteractionMode.Read, resolveMessageMode(FlareMessageLifecycle(read = FlareMessageReadState.Read)))
    }

    @Test fun `selection shortcuts and gestures are stable`() {
        val anchor = reduceSelection(FlareSelectionState(), FlareSelectionReplace("b"))
        val range = reduceSelection(anchor, FlareSelectionExtend("d", listOf("a", "b", "c", "d")))
        assertEquals(setOf("b", "c", "d"), range.selectedIds)
        assertEquals(FlareDesktopShortcutAction.CommandPalette, resolveDesktopShortcut("k", primary = true))
        assertEquals(FlareSwipeIntent.None, resolveSwipeIntent(80f, 70f, message = true))
        assertEquals(FlareSwipeIntent.Reply, resolveSwipeIntent(80f, 8f, message = true))
    }

    @Test fun `message capabilities follow ownership and terminal lifecycle`() {
        val own = resolveMessageCapabilities(FlareMessageCapabilityInput(FlareMessageLifecycle(), own = true))
        assert(own.contains(FlareMessageCapability.Edit))
        assertFalse(own.contains(FlareMessageCapability.Report))
        val recalled = resolveMessageCapabilities(
            FlareMessageCapabilityInput(
                FlareMessageLifecycle(mutation = FlareMessageMutationState.Recalled),
                own = true,
            ),
        )
        assertEquals(emptySet<FlareMessageCapability>(), recalled)
    }

    @Test fun `extended actions and shortcuts share the capability contract`() {
        val actions = resolveMessageCapabilities(
            FlareMessageCapabilityInput(
                FlareMessageLifecycle(), own = true, pinned = true,
                hasThread = true, hasQuote = true, supportsMergeForward = true,
            ),
        )
        assert(actions.containsAll(setOf(
            FlareMessageCapability.Unpin, FlareMessageCapability.OpenThread,
            FlareMessageCapability.JumpToQuote, FlareMessageCapability.MergeForward,
        )))
        assertEquals(FlareDesktopShortcutAction.Forward, resolveDesktopShortcut("f", scope = FlareDesktopShortcutScope.Conversation))
        assertEquals(FlareDesktopShortcutAction.OpenThread, resolveDesktopShortcut("t", scope = FlareDesktopShortcutScope.Conversation))
    }

    @Test fun `action presentation stays downstream of availability`() {
        val actions = resolveMessageActions(
            FlareMessageCapabilityInput(FlareMessageLifecycle(), own = false),
            FlareMessageActionPresentation.HoverToolbar,
        )
        assert(actions.single { it.id == FlareMessageCapability.Reply }.promoted)
        assertEquals(FlareMessageActionGroup.Destructive, actions.single { it.id == FlareMessageCapability.Report }.group)
    }
}
